src/java.base/share/classes/com/sun/crypto/provider/PBES2Core.java

```diff
@@ -177,96 +195,70 @@ protected void engineInit(int opmode, Key key,
 
-            // TBD: consolidate the salt, ic and IV parameter checks below
-
-            // Extract salt and iteration count from the key, if present
-            if (key instanceof javax.crypto.interfaces.PBEKey) {
-                salt = ((javax.crypto.interfaces.PBEKey)key).getSalt();
-                if (salt != null && salt.length < 8) {
+            boolean doEncrypt = ((opmode == Cipher.ENCRYPT_MODE) ||
+                        (opmode == Cipher.WRAP_MODE));
+
+            // Extract from the supplied PBE params, if present
+            if (params instanceof PBEParameterSpec pbeParams) {
+                // salt should be non-null per PBEParameterSpec
+                salt = check(pbeParams.getSalt());
+                iCount = check(pbeParams.getIterationCount());
+                AlgorithmParameterSpec ivParams = pbeParams.getParameterSpec();
+                if (ivParams instanceof IvParameterSpec iv) {
+                    ivSpec = iv;
+                } else if (ivParams == null && doEncrypt) {
+                    // generate random IV
+                    byte[] ivBytes = new byte[blkSize];
+                    random.nextBytes(ivBytes);
+                    ivSpec = new IvParameterSpec(ivBytes);
+                } else {
                     throw new InvalidAlgorithmParameterException(
-                            "Salt must be at least 8 bytes long");
+                            "Wrong parameter type: IV expected");
                 }
-                iCount = ((javax.crypto.interfaces.PBEKey)key).getIterationCount();
-                if (iCount == 0) {
-                    iCount = DEFAULT_COUNT;
-                } else if (iCount < 0) {
-                    throw new InvalidAlgorithmParameterException(
-                            "Iteration count must be a positive number");
+            } else if (params == null && doEncrypt) {
+                // Try extracting from the key if present. If unspecified,
+                // PBEKey returns null and 0 respectively.
+                if (key instanceof javax.crypto.interfaces.PBEKey pbeKey) {
+                    salt = check(pbeKey.getSalt());
+                    iCount = check(pbeKey.getIterationCount());
                 }
-            }
-
-            // Extract salt, iteration count and IV from the params, if present
-            if (params == null) {
                 if (salt == null) {
-                    // generate random salt and use default iteration count
+                    // generate random salt
                     salt = new byte[DEFAULT_SALT_LENGTH];
                     random.nextBytes(salt);
+                }
+                if (iCount == 0) {
+                    // use default iteration count
                     iCount = DEFAULT_COUNT;
                 }
-                if ((opmode == Cipher.ENCRYPT_MODE) ||
-                        (opmode == Cipher.WRAP_MODE)) {
+                //if (ivSpec == null) { // old behavior always generate
```

Comment:
How could `ivSpec` be non-null here? IIUC the only answer is from a previous `engineInit`, and it should not be retained. I suggest removing this check plus adding an explicit `ivSpec = null` at the beginning of this method.

Reply:
Yeah, it's also possible that ivSpec is non-null if getParameters() is called before init(). Now that salt and iCount is reset in the beginning of init(), ivSpec should be reset too.

Reply:
Oh, I didn't realize that. So now whenever `init()` is called every old param (no matter why it was set) is totally wiped. I like this consistent behavior.